### PR TITLE
nixos-render-docs: init redirects system

### PR DESCRIPTION
**Pull request by GetPsyched**
An implementation of a redirects system to keep track of changes in the documentation such that old links don't break. Such a system will enable free movement of pages in the future without fear of breaking links; while also allowing us to catch old breakages retrospectively -- in a follow-up PR :D

PS: Commits will be squashed once the PR is marked for review.

### CODE REVIEW

File: pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:89]:
```suggestion
        """Test moving a leaf identifier to a different output file."""
```

This is done implicitly through how the boilerplate is set up, so it's not enough to say it's just moving to a different source file.

File: pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:178]:
```suggestion
        expected_script = f"const redirects = {json.dumps(expected_redirects)};"
        self.assertEqual(result, expected_script)
```

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py:116]: TODO for later: compute the relative path between `redirection_target` (maybe a better name: `from`?) and `locations[0]`

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py:118]:
```suggestion
        return self._redirects_script.replace('REDIRECTS_PLACEHOLDER', json.dumps(client_redirects), sort_keys=True)
```

We should keep the order predictable, otherwise we'll surely run into issues later on. Right now it doesn't show because we're reading in file order and writing in processing order, but who knows how this code will evolve.

File: pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:161]: Either `sort_keys=True` or read back the JSON into a dict. Or even better: let `get_client_redirects()` return JSON, and serialize that to a string at the call site.

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py
Comment (GetPsyched) [reply pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py]: Done this for the most part except:
> Update the exception subclass accordingly, and move the string generation functions together in there.

Could you clarify the goal? Is it to merge all of the exception classes into just one `RedirectsError` class that outputs a string dependent on all the collections passed to it? That sounds like a good idea, but my concern with that is not being able to assert specific errors in the tests.

File: pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py
Comment (GetPsyched) [reply pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:200]: My concern with having `get_client_redirects()` return a dict is that it moves the responsibility of handling the substitution to `_file_header()` and it isn't really its job to do that? Or am I wrong to think that

Comment (fricklerhandwerk) [reply pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:200]: Why not? Why should the redirects be a data structure *baked into a string* instead of the actual data structure, at any point except where you absolutely need a string? The string should be the final representation that we don't touch any more, IMO.

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py
Comment (fricklerhandwerk) [reply pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py]: I thought about something like

```python
def test_validation_catches_specific_condition():
    thing = Thing()
    thing.do_stuff() 
    
    with pytest.raises(RedirectsError) as exc_info:
        thing.validate()
    
    assert exc_info.value.conflicting_anchors != None
```

File: pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py
Comment (GetPsyched) [reply pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:200]: Fair enough.

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py
Comment (GetPsyched) [reply pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py]: Hmm, that looks about right, I'll implement this.

**Review by fricklerhandwerk — APPROVED** [review]
This looks good architecturally. Tests and build still fail though.

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py:17]: This is a dict elsewhere

File: pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:190]: I wonder how this an pass, because [here](https://github.com/NixOS/nixpkgs/pull/353513/files#diff-9bdf5c28b6cdf553d0d4e6735b57327691c4ad1bb8306e036c2b213fc460abe8R176) we're assigning dict fields

Comment (GetPsyched) [reply pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:190]: Unsure, but I remember the unit test logs showing a list, not a dict -- again, not sure how that's possible; but I'll fix the types.

File: pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py
Comment (fricklerhandwerk) [reply pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:188]: Also I don't think that with the redirects checks enabled, we'll ever have a situation where a file will move as a whole merely by moving identifiers. Even top-level headings will require creating such an element-wise entry.

Comment (GetPsyched) [reply pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:188]: > What we'd want in that situation is a new server-side redirect bar.html -> foo.html and that requires rewriting all bar.html entries in the redirects database to foo.html because otherwise those redirects just won't apply for the new output file.

That's precisely what I'm enforcing with this rule.

> We could catch such a situation programmatically, in the special case where the elements in the new location are a superset of the elements in the old location.

That's certainly possible, with little effort too I think. It should be as simple as modifying `get_client_redirects()` to check if its `redirection_target` exists as a destination path of some server-side redirect, and pull in all the client-side redirects from its the source path.

Should we do it this way? It would lead to smaller diffs when server-side redirects ever happen. Not sure what are the cons of doing this.

Comment (fricklerhandwerk) [reply pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py:188]: > That's certainly possible, with little effort too I think. It should be as simple as modifying get_client_redirects() to check if its redirection_target exists as a destination path of some server-side redirect, and pull in all the client-side redirects that originate from its the source path.

Yes, sounds about right. Again, this is for a follow-up.

**Comment by hsjobeki** [comment]
> Do we have any options to not commit a 70k lines json? 😅

It seems like the content is trivial too.
Could it be possible to compute the content as good as possible during the build if this file is needed. 
I guess it is there for maintaining custom redirects. But it should be possible to only maintain what actually needs to be customized.

**Comment by hsjobeki** [comment]
@GetPsyched Thank you so much for doing this.

Once the pages are split i would like to integrate pagefind. This should be straight forward. 
We should also export a sitemap xml i'll look into that so we also get better discoverability of our content by google.

**Comment by GetPsyched** [comment]
> > Do we have any options to not commit a 70k lines json? 😅
> 
> It seems like the content is trivial too. Could it be possible to compute the content as good as possible during the build if this file is needed. I guess it is there for maintaining custom redirects. But it should be possible to only maintain what actually needs to be customized.

I don't love that we're adding huge files when nixpkgs size is a growing concern; this is why a lot of effort went into making sure we have the most minimum diff.

We're compelled to storing all identifiers as its static data we cannot infer in the NRD's build process itself. One way to do it would be to check the base/parent commit, but that has its own issues; some that I can remember off the top of my head:
- Non-trivial overhead to contributors who'll have to run external command(s) after each change.
- Unsure how feasible is it to develop a CI check for this; and if there isn't one, contributors and reviewers are certainly going to forget to run such a command if there's no error to yell at them.

An alternative that was considered was storing each identifier as a file instead of a key in a JSON. We chose not to go that route as it would lead to having 20,000+ files as of 24.05.

**Comment by GetPsyched** [comment]
> @GetPsyched Thank you so much for doing this.
> 
> Once the pages are split i would like to integrate pagefind. This should be straight forward. We should also export a sitemap xml i'll look into that so we also get better discoverability of our content by google.

Awesome! I've heard good things about pagefind and wanted to try it out myself at some point. IIRC it works off of the rendered HTML, right? This should fit quite well with the manuals.

I'll get to the XML bit once a few key things are checked off our todo; will ping you on the documentation Matrix room :D

File: pkgs/tools/nix/nixos-render-docs/README.md
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/README.md:11]:
```suggestion
Maintaining our own documentation rendering framework may appear extreme but has practical advantages:
- We never have to work around existing tools made under different assumptions
- We don't have to deal with unexpected breakage
- We can grow the framework with our evolving requirements without relying on external support or approval or the need to maintain a small diff to upstream
- The amount of code involved is minimal because it's single-purpose
```

File: pkgs/tools/nix/nixos-render-docs/README.md
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/README.md:20]:
```suggestion
The [mechanism for checking correctness of redirects](./src/nixos_render_docs/redirects.py) takes the collection of identifiers and a mapping of the identified content to its historical locations in the output.
It validates them against a set of rules, and creates a client-side redirect mapping for each output file, as well as a `_redirects` file for server-side redirects in [Netlify syntax](https://docs.netlify.com/routing/redirects/#syntax-for-the-redirects-file).
```

File: pkgs/tools/nix/nixos-render-docs/README.md
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/README.md:29]:
```suggestion
The creation, movement, and removal of every identifier is captured in the Git history.
```

File: pkgs/tools/nix/nixos-render-docs/README.md
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/README.md:38]:
```suggestion
However, analysing hundreds of thousands of commits during the build process is impractical.
```

Let's not go into vagueness about alternatives considered. Either spell them out or leave them out.

File: pkgs/tools/nix/nixos-render-docs/README.md
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/README.md:44]:
```suggestion
    - It would also require keeping an impure or otherwise continuously updated reference to those other revisions.
```

File: pkgs/tools/nix/nixos-render-docs/README.md
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/README.md:47]:
```suggestion
    - The static mapping acts like a semi-automatically updated cache that we drag along with version history.
    - Other setups, such as a dedicated service to cache a history of moved content, are more complicated and would still be impure.
- Checking in large amounts of data that is touched often, bears a risk of more merge conflicts or related build failures.
```

File: pkgs/tools/nix/nixos-render-docs/README.md
Comment (fricklerhandwerk) [inline pkgs/tools/nix/nixos-render-docs/README.md:50]:
```suggestion
The solution picked here is to have a static mapping of the historical locations checked into the Git tree, such that it can be read during the build process.
```

File: pkgs/tools/nix/nixos-render-docs/README.md
Comment (GetPsyched) [reply pkgs/tools/nix/nixos-render-docs/README.md]: But it _was_ lossy, as written in the RFC.

**Comment by hsjobeki** [comment]
> > > Do we have any options to not commit a 70k lines json? 😅
> > 
> > It seems like the content is trivial too. Could it be possible to compute the content as good as possible during the build if this file is needed. I guess it is there for maintaining custom redirects. But it should be possible to only maintain what actually needs to be customized.
> 
> I don't love that we're adding huge files when nixpkgs size is a growing concern; this is why a lot of effort went into making sure we have the most minimum diff.
> 
> We're compelled to storing all identifiers as its static data we cannot infer in the NRD's build process itself. One way to do it would be to check the base/parent commit, but that has its own issues; some that I can remember off the top of my head:
> - Non-trivial overhead to contributors who'll have to run external command(s) after each change.
> - Unsure how feasible is it to develop a CI check for this; and if there isn't one, contributors and reviewers are certainly going to forget to run such a command if there's no error to yell at them.
> 
> An alternative that was considered was storing each identifier as a file instead of a key in a JSON. We chose not to go that route as it would lead to having 20,000+ files as of 24.05.

I want to look into those details more closely. I am concerned about the size of the File. And huge diffs cluttering git. Please give me some time i ll do it then next days and provide alternatives If i find any

**Comment by fricklerhandwerk** [comment]
@hsjobeki: @GetPsyched  and I discussed a bunch of alternatives, and I'd be happy to get your input on it, too.

In a sense, keeping the checked-in file is a maximum-fidelity maximum-speed solution at the cost of file size and risk of merge conflicts. One possible trade for Git repo size against fidelity would be only guaranteeing stable URLs against the past two stable releases -- but that comes at the cost build time (as we'd have to fetch the release branches and run `nixos-render-docs` to extract identifiers and locations) and the requirement to keep the pointers to the release branches up to date. While we can make the checks optional and only enforce them in CI, we'd increase the contributor feedback cycle by a great deal.

I'd be glad if we found a solution that does the important things well.

File: pkgs/tools/nix/nixos-render-docs/README.md
Comment (fricklerhandwerk) [reply pkgs/tools/nix/nixos-render-docs/README.md]: This is only half-true. The conversion was definitely loss for the Nix manual, and also broke tons of URLs and workflows. `nixos-render-docs` was tested against the DocBook output and did not loose output fidelity.

Comment (GetPsyched) [reply pkgs/tools/nix/nixos-render-docs/README.md]: Ah, okay. I have limited knowledge on this so I simply flagged it in case it was an error.

File: pkgs/by-name/ni/nixos-render-docs/src/nixos_render_docs/manual.py
Comment (hsjobeki) [inline pkgs/by-name/ni/nixos-render-docs/src/nixos_render_docs/manual.py:324]: This embeds `2.3M` of text into the index.html. Potentially more in the future, when people start to extend the map file.

Better:

`<script type="module" src="path/to/script.js"></script>` (generate a file, instead of inlining 2.3 MB, which is pretty bad for loading time)

This allows the browser to load the script seperate from the content. Javascript-modules are supported across all major browser. This also gives us the possibility in the future to only load the file if we detect a user coming from an old anchor.

**Comment by hsjobeki** [comment]
Some feedback and suggestions on the redirects.json.

Conceptually the format provides very good stability of links. Probably more than is needed.
I am thinking about a compromise for both docs maintainers and ux for readers.

- Options from the options.html (20k) produce 70k lines of redirects.json (~2.3MB). The options documentation is automatically generated. Content is absolutely identical with search.nixos.org/options. Therefore I dont see the benefit of making those references stable. How is the option documentation split into pages?
Since all these references are subtle to break also with renaming / moving the options itself this would introduce a duplicate source of truth?

- References tracking for the nixpkgs manual: 
Maybe we can limit backwards compatibility by only providing stability down to h3 headings. (Chap, Section, sub-section, i.e. 3 levels)

For example we declare this as stable and maintained.
`index.html#module-system-lib-evalModules`

But 
`index.html#module-system-lib-evalModules-return-value-options`
This is not stable and is cut-off / remapped to `index.html#module-system-lib-evalModules`
 which is the closest maintained redirect. Once we split the pages the visitor would end up on the correct sub page. but needs to scroll down to the `return-value-options` manually.
 
 Seems like a good compromise to me.

File: pkgs/by-name/ni/nixos-render-docs/src/nixos_render_docs/manual.py
Comment (GetPsyched) [reply pkgs/by-name/ni/nixos-render-docs/src/nixos_render_docs/manual.py:324]: It does not embed `2.3M` of redirects into `index.html`. We only send in redirects that actually exist, not what you see in `redirects.json`. See [`get_client_redirects()`](https://github.com/GetPsyched/nixpkgs/blob/dca60b3027d630732e2670b92b7cc13622c0939b/pkgs/by-name/ni/nixos-render-docs/src/nixos_render_docs/redirects.py#L143-L154)

That said, I agree with you on the inline vs. module JS suggestion, I will make that change.

**Comment by fricklerhandwerk** [comment]
@Hsjobeki thanks for the additional context, this makes a lot of sense.

Just as a side note: tracking renames of options and libraries boils down to the same problem we're trying to slow there. Ultimately we somehow want to make sure that a *diff* has the right semantics, e.g. "never remove an option". But all we can *typically* do is check that any given current state of the repo has the right structure. For better correctness guarantees we probably need both.

One thing that we had already discussed and found a satisfactory answer for: when we remove things, their identifiers should redirect to the release note that announces the removal and hopefully explains why.

@getpsyched then next step would be grabbing the identifiers from the option and library documentation, and filtering out those and the auto-ids for the static check. And the source based identifiers need to be somehow evaluated to derive redirects from them.

@Hsjobeki we plan to add tools to make the diffs that would fix errors that are currently only reported. In most cases the issues should be narrow enough to have simple but reliable automation.

**Comment by fricklerhandwerk** [comment]
@hsjobeki do you think we should also block merging on the tools to fix the issues automatically? Given we reduce the scope of the checks to the hand-written identifiers, not that many people should be affected by the new constraints. Which is why I would suggest not blocking on the helpers, but only on filtering out the automatic identifiers from the static mapping.

**Comment by hsjobeki** [comment]
> If this gets too involved, the very least we could do is filtering out the library identifiers entirely and just not making any new guarantees for now. Actually, we could unblock this PR by merely filtering out all auto-identifiers and then adding redirect construction for them later.

That is what i would suggest.

I guess once the pages are split out the new automatic library and option docs will follow a certain schema.
The migration for those can be done alongside with splitting them.

Assuming that lib functions and options are stable. This will also yield stable redirects.

For breaking changes of the underlying references (lib, options) i would suggest that we introduce another  kind of index for documentation that documents lib functions that have been deprecated alongside with the reason and some other stuff.

The deprecation index for options should be possible to build from the options themselves somehow. But i would need to look into that details (option tree) a bit further.

File: pkgs/by-name/ni/nixos-render-docs/src/nixos_render_docs/redirects.py
Comment (fricklerhandwerk) [inline pkgs/by-name/ni/nixos-render-docs/src/nixos_render_docs/redirects.py:92]:
```suggestion
            # filter out automatically generated identifiers from module options and library documentation
            if id.startswith("opt-") or id.startswith("auto-generated-"):
```

Comment (fricklerhandwerk) [reply pkgs/by-name/ni/nixos-render-docs/src/nixos_render_docs/redirects.py:93]:
```suggestion
            if id.startswith("opt-") or id.startswith("auto-generated-") or id.startswith("function-library-"):
```

We probably want to catch those, too. Not sure if this removes too much, because I can't grep out of the code if there are such identifiers created manually. Very likely not, since it's [created by nixdoc](https://github.com/nix-community/nixdoc/blob/4fa541a7b270a255caf6bbceeb29d919e6e66704/src/commonmark.rs#L164), but we have to hope that noone wanted to be clever.

**Comment by hsjobeki** [comment]
> @hsjobeki we're down to ~6400LOC for tracking redirects on hand-written content. Any objections to merging?

Seems as low as it may get for now. I am okay with merging as it is.

Could you add me as codeowner as well? So i get notified if anyone starts actively changing/adding redirects. 

Also i feel like we are missing some documentation how to test a redirect after adding it. (nixos/nixpkgs manual)

File: ci/OWNERS
Comment (hsjobeki) [inline ci/OWNERS:111]:
```suggestion
# NixOS Render Docs
/pkgs/by-name/ni/nixos-render-docs @fricklerhandwerk @GetPsyched @hsjobeki
/doc/redirects.json                @fricklerhandwerk @GetPsyched @hsjobeki
/nixos/doc/manual/redirects.json   @fricklerhandwerk @GetPsyched @hsjobeki
```

**Comment by fricklerhandwerk** [comment]
Something like: `cd doc; nix-build; firefox --new-tab file://$(pwd)/result/index.html#my-anchor` and observe that it redirects

For server-side redirects you'd need to run `netlify-cli` like we do on nix.dev: https://github.com/NixOS/nix.dev/blob/master/CONTRIBUTING.md#local-preview

**Comment by GetPsyched** [comment]
NixOS and Nixpkgs manual CI is failing due to a lack of a rebase. I will do it once the changes have been reviewed.

**Review by fricklerhandwerk — APPROVED** [review]

**Review by hsjobeki — APPROVED** [review]

**Comment by GetPsyched** [comment]
I think we need to backport this PR since branch-off happened yesterday.
cc @fricklerhandwerk @hsjobeki

**Pull request merged**
